components/callbacks/statemachine.go

```diff
@@ -113,6 +114,27 @@ func (c Callback) output() (hsm.TransitionOutput, error) {
 	return hsm.TransitionOutput{Tasks: tasks}, err
 }
 
+// TODO: Remove this implementation once transition history is fully implemented.
+func (c Callback) progress() (int, int32, error) {
+	switch c.State() {
+	case enumsspb.CALLBACK_STATE_UNSPECIFIED:
+		return 0, 0, serviceerror.NewInvalidArgument("uninitialized callback state")
+	case enumsspb.CALLBACK_STATE_STANDBY:
+		return 1, 0, nil
+	case enumsspb.CALLBACK_STATE_SCHEDULED:
+		return 2, c.GetAttempt() * 2, nil
+	case enumsspb.CALLBACK_STATE_BACKING_OFF:
+		// We've made slightly more progress if we transitioned from scheduled to backing off.
+		return 2, c.GetAttempt()*2 + 1, nil
+	case enumsspb.CALLBACK_STATE_FAILED, enumsspb.CALLBACK_STATE_SUCCEEDED:
+		// Consider any terminal state as "max progress", we'll rely on last update namespace failover version to break
```

Comment:
Synced offline. I personally feel CompareState should return an error if two states are non-comparable to make the implementation more self-contained. Not critical though.